contracts/nftauction.js

```diff
@@ -0,0 +1,597 @@
+    && el.data.symbol === symbol) !== undefined) {
+    return true;
+  }
+
+  return false;
+};
+
+const sendNfts = async (to, wrappedNfts) => {
+  // transfer NFTs
+  await api.executeSmartContract('nft', 'transfer', {
+    fromType: 'contract',
+    to,
+    toType: 'user',
+    nfts: [wrappedNfts],
+    isSignedWithActiveKey: true,
+  });
+};
+
+const returnBids = async (bids, priceSymbol) => {
+  for (let i = 0; i < bids.length; i += 1) {
+    const {
+      account,
+      bid,
+    } = bids[i];
+
+    // send tokens back to bidders
+    await api.transferTokens(account, priceSymbol, bid, 'user');
+  }
+};
+
+const cancelAuction = async (auction) => {
+  const {
+    auctionId,
+    symbol,
+    seller,
+    nftIds,
+    priceSymbol,
+    bids,
+  } = auction;
+
+  // return all the bids
+  await returnBids(bids, priceSymbol);
+
+  const wrappedNfts = {
+    symbol,
+    ids: nftIds,
+  };
+
+  // return NFTs to the seller
+  await sendNfts(seller, wrappedNfts);
+
+  api.emit('cancelAuction', {
+    auctionId,
+  });
+};
+
+const settleAuction = async (auction, id = null) => {
+  const {
+    auctionId,
+    symbol,
+    seller,
+    nftIds,
+    priceSymbol,
+    bids,
+    currentLead,
+  } = auction;
+
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestamp = blockDate.getTime();
+
+  const wrappedNfts = {
+    symbol,
+    ids: nftIds,
+  };
+
+  if (bids.length > 0) {
+    const bidId = id === null ? currentLead : id;
+    const leadBid = bids[bidId];
+
+    // remove the lead bid from the returning bids
+    bids.splice(bidId, 1);
+
+    // return the bids
+    await returnBids(bids, priceSymbol);
+
+    // send payment to the seller
+    await api.transferTokens(seller, priceSymbol, leadBid.bid, 'user');
+
+    // send NFTs to the buyer
+    await sendNfts(leadBid.account, wrappedNfts);
+
+    await api.db.remove('auctions', auction);
+
+    api.emit('settleAuction', {
+      auctionId,
+      symbol,
+      seller,
+      nftIds,
+      bidder: leadBid.account,
+      price: leadBid.bid,
+      priceSymbol,
+      timestamp,
+    });
+  } else {
+    // if there are no bids, expire the auction
+    await api.db.remove('auctions', auction);
+
+    // return NFTs to the seller
+    await sendNfts(seller, wrappedNfts);
+
+    api.emit('expireAuction', {
+      auctionId,
+      symbol,
+      seller,
+      nftIds,
+      timestamp,
+    });
+  }
+};
+
+actions.create = async (payload) => {
+  const {
+    symbol,
+    nfts,
+    minBid,
+    finalPrice,
+    priceSymbol,
+    expiry,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(symbol && typeof symbol === 'string', 'invalid symbol')) return;
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')) return;
+
+  const nft = await api.db.findOneInTable('nft', 'nfts', { symbol });
+
+  if (api.assert(nfts && typeof nfts === 'object' && Array.isArray(nfts)
```

Comment:
here and below: i think can be cleaner if we use early exit as in line 229:  if (!api.assert(...)) return